### DISCUSSION TITLE
CI with GitHub Action

### DISCUSSION THREAD

**Discussion by tiborauer** [automaticanalysis/automaticanalysis#272] [Show and tell]
I have implemented CI for aa using GitHub Action. It works really great and allows automatic testing. It runs all four current test scripts in parallel.
[Here](https://github.com/automaticanalysis/automaticanalysis/tree/master/.github/workflows) you can find the corressponding files, and [here](https://github.com/automaticanalysis/automaticanalysis/actions) you can take a look at [the most recent run](https://github.com/automaticanalysis/automaticanalysis/actions/runs/4054307634) and even download the artifacts.

As you can see there are two (minor) issues:
-  FaceMasking has issues with libtinfo.so.6 and triggers an error "no version information available". It must be environment-dependent because I could not replicate it on our system. After downloading the artifacts, it seems to be related to MATLAB licensing. FaceMasking tool is a shell-based tool calling MATLAB, and the "MATLAB-calling-shell-calling-MATLAB" scenario might further confound the case for GitHub-hosted runners.
- The disk space on the GitHub-hosted runners is limited and not enough for the current fMRI test with 4 subjects (required by inferencing with FWE) and two ssessions. We have two options: either reduce the subject number (and replace FWE with "none") or reduce the number of sessions (and omit branching). Which one would you prefer?

**Comment by AljenU** [comment]
Why is it that the full results are uploaded as artifacts?

**Reply by tiborauer** [reply]
Without #244 in place, it is a bit tedious to filter through all derived data. Besides, it allows a more in-depth investigation of the results.
@jones-michael-s, wrote [a jpeg collector](https://github.com/automaticanalysis/automaticanalysis/blob/master/developer/aa_jpeg_crawler.m), but I did not have time to put it into practice.

**Reply by AljenU** [reply]
In general, the console output will tell if something is wrong, and what. We do not even have a reference currently for looking at the produced output and checking that it is what is expected. Thus, from my perspective, it only adds extra runtime to the jobs (for the fMRI, 25 minutes). And uploading only a single file that we expect to have been produced would give the same confirmation that the job has finished 'correctly'.

**Comment by AljenU** [comment]
It looks like aa_test is called with haltonerror with default value, i.e. false. So in case of an error, the action gets to a keyboard command and stops there waiting for input?

**Reply by tiborauer** [reply]
That would be the case only if `aap.options.wheretoprocess ~= 'localsingle'`.

**Reply by AljenU** [reply]
Ah, yes, so it does continue. Since the error is caught, doesn't the matlab session finish with a regular success exit code, and thus the CI does then not even see that there was an error, and will report success?

**Reply by tiborauer** [reply]
True. It just occurred to me that since we run the tests separately, it makes more sense to set `haltonerror = 1`.

**Reply by AljenU** [reply]
That will probably be the easiest first fix.

Another option would be to use a wrapper / rewrite aa_test to use directly, the matlab unittest framework. That can give details on why things failed, directly in the github actions interface.

**Reply by tiborauer** [reply]
The Matlab unittest framework seems more efficient if we run multiple tests at once. Unfortunate;y, it is not an option because there is not enough space on GitHub-hosted runner. Besides, we can now run all tests in parallel.

**Reply by tiborauer** [reply]
OK. I have changed 'haltonerror'

**Comment by AljenU** [comment]
A lot of code running shell commands is duplicated in the 4 jobs. Which means, that if you want to change something (e..g. how SPM is downloaded), it needs to be changed 4 times. It would be better to have a shell script for each individual action, as part of the repository. Then as soon as the repository has been checked out, can then call the shell scripts. And only that one script needs to be changed.

I would then just have one script per dependency (FSL, Freesurfer etc), and one action for each. Currently downloading the dependencies is one step.

**Reply by AljenU** [reply]
Actually, there could be just 1 wrapper shell script, that calls the others. With environment variables determining what dependency-loading scripts are ran. You would still set the AWS env vars from the .yml, like e.g. AWS_EC2_METADATA_DISABLED. Similarly there could be env vars for LOAD_FSL and LOAD_FREESURFER.
Do note that the shell scripts should fully exit on error, and not continue with the next sub-shellscript.

**Reply by tiborauer** [reply]
I think it is an excellent suggestion! Moreover, the users could also use such script(s) to install (a minimum set of) dependencies. Of course, the script(s) will be OS-specific (the CI, for example, uses ubuntu), therefore, we have to come up with more of them.

**Reply by tiborauer** [reply]
I am on it.

**Reply by AljenU** [reply]
I think we should be careful in wanting to have the scripts be dual-use, that is, both for use on the github runner and by regular users. Since the scripts are inter-dependent, and thus have quite a lot of assumptions. Also, the contents of the _github parameter xml.

For instance, on the github runner, awscli comes preinstalled (so we actually do not need "sudo apt-get install awscli" in the jobs). Whereas for user install, that assumption cannot be made.

**Reply by tiborauer** [reply]
I have updated the tool installation framework. I wrote some comments which - hopefully - makes it useful for other systems, as well. The corresponding code is in 
- .github/workflows
- aa_tools/toolboxes/installation_scripts

**Comment by AljenU** [comment]
In the action install tools, wget and unzip are run with the -q modifier. However, this makes it harder to see where things failed if they do (like currently with all the exit code 4 errors. And, you only go look at the output when someting goes wrong, so it's not bad if there is a bit more.

**Reply by tiborauer** [reply]
I figured out that the download of FaceMasking is failing. There is something with the hosting site. I contacted the developer and they are investigating the issue.

**Reply by tiborauer** [reply]
It has been sorted now.

**Reply by AljenU** [reply]
Thanks. Very nice to have some CI in place, it gives at least some idea that nothing fundamental broke when a PR is submitted.

**Reply by tiborauer** [reply]
I have decided to remove FaceMasking from the mix until we can sort it out.

**Comment by tiborauer** [comment]
@AljenU and @jones-michael-s 
Tee CI is now operable again, and it runs through correctly. Please, fetch and merge the central repo into your repos and PRs.

**Comment by AljenU** [comment]
Running on Mac and Windows:

Adding a mac runner should probably work quite ok already? Though there may be parts of the installation scripts that need to be adapted, so it will be needed to be tested. But at least the tests themselves should run ok once the installation scripts are there.

I do want a Windows job as well, however on Windows the main integration tests that are now the official tests will not work yet, and not sure what it will take to get them running. I do want tests that the Windows changes that are made, are not broken again. Thus i think it would be good to have some more tests, that work on all platforms. For that, i do want to use the matlab unittest framework as a wrapper, since it has features that make it easy to select tests, a.o. based on tags, so we can have WindowsOnly tests for instance.

**Reply by tiborauer** [reply]
I am afraid I am not a mac person and not able to set up a mac installation.

**Comment by tiborauer** [comment]
It seems the AROMA use case test is sometimes killed, and I cannot retrieve more debug information. Redoing the test, however, make it pass eventually.
It is just something you should keep in mind if you see the AROMA use case test fail.

**Comment by jones-michael-s** [comment]
Is killing the test case indicative that AROMA crashed? 

Our experience is AROMA may crash if it can't denoise a dataset. That's doesn't quite sound like the whole story here since (if I'm understanding correctly) sometimes the test passes and it's always the same data. But I vaguely remember AROMA does some random initialization (although I could be remembering wrong). Maybe that occasionally fails?

**Reply by tiborauer** [reply]
Yes. It is a random fail because it eventually passes for the second or third time.
